feat: simplified invocation context

psychedelicious opened the pull request:
## What type of PR is this? (check all applicable)

- [x] Refactor
- [x] Feature
- [ ] Bug Fix
- [x] Optimization
- [ ] Documentation Update
- [ ] Community Node Submission


## Have you discussed this change with the InvokeAI team?
- [x] Yes
- [ ] No, because:

      
## Have you updated all relevant documentation?
- [x] Yes
- [ ] No


## Description

This PR iterates on the rough draft in #4910, taking into consideration feedback from @RyanJDick. It's written from scratch.

I'm sure there are a few methods that node authors will want that are not yet included - I imagine we'll need more boards-related methods. The new context is easily extensible.

The app ~~is~~ should be fully functional in this PR. I've converted everything.

See `invocation_context.py` for the meat and potatoes. It's well-commented.

There are a crapload of changes, of course - every invocation needed to be updated. And I had to shuffle files around to deal with circular imports.

Some of the feedback from the earlier PR was that it would be nice to redo the `BaseInvocation` API at the same time. We should have some discussion around that. Maybe we can get a revised invocation context (this PR or a different design if this doesn't work out) and revised `BaseInvocation` for v3.7.0.

## Related Tickets & Documents

<!--
For pull requests that relate or close an issue, please include them
below. 

For example having the text: "closes #1234" would connect the current pull
request to issue 1234.  And when we merge the pull request, Github will
automatically close the issue.
-->

- Closes #4887
- Supersedes #4910

## QA Instructions, Screenshots, Recordings

Note that I have not touched `onnx.py`. It's been broken for a while so I'm just ignoring it.

The app should run just fine.

<!-- 
Please provide steps on how to test changes, any hardware or 
software specifications as well as any other pertinent information. 
-->

## Merge Plan

This PR needs thorough review, testing and pondering. If it is approved, I will handle the actual merging.

<!--
A merge plan describes how this PR should be handled after it is approved.

Example merge plans:
- "This PR can be merged when approved"
- "This must be squash-merged when approved"
- "DO NOT MERGE - I will rebase and tidy commits before merging"
- "#dev-chat on discord needs to be advised of this change when it is merged"

A merge plan is particularly important for large PRs or PRs that touch the
database in any way.
-->

## Added/updated tests?

- [x] Yes
- [ ] No : _please replace this line with details on why tests
      have not been included_

## [optional] Are there any post deployment tasks we need to perform?

This will be a breaking change for node authors. We'll need to provide solid documentation.

psychedelicious commented:
I've added the old `InvocationContext` attrs to the new version and added deprecation warnings.

While this makes it easier for custom nodes to migrate, the file structure changes needed to avoid circular references mean that custom nodes need to be revised anyways, so I'm not sure it makes sense to handle deprecation.

I've added an `invocation_api` module (thanks for this suggestion Ryan!) at `invokeai.invocation_api` and added this to the package. Custom nodes can import everything they may need from this file. There may be some classes I've missed in this file.

brandonrising commented on invokeai/app/invocations/onnx.py:
```diff
@@ -316,7 +318,7 @@ def dispatch_progress(
         name = f"{context.graph_execution_state_id}__{self.id}"
         context.services.latents.save(name, latents)
-        return build_latents_output(latents_name=name, latents=torch.from_numpy(latents))
+        # return build_latents_output(latents_name=name, latents=torch.from_numpy(latents))
```
oh is it related to this in your description 

> Note that I have not touched onnx.py. It's been broken for a while so I'm just ignoring it.

psychedelicious replied:
Yep. Added a comment to the file clarifying.

brandonrising replied:
almost feels like we should just remove this file as part of this

brandonrising replied:
Would cause some issues in the front end though

psychedelicious commented on invokeai/app/invocations/onnx.py:
I only updated this file enough so that the app runs.

brandonrising commented on invokeai/app/invocations/compel.py:
```diff
@@ -66,25 +64,17 @@ class CompelInvocation(BaseInvocation):
             return
 
-        # loras = [(context.services.model_manager.get_model(**lora.dict(exclude={"weight"})).context.model, lora.weight) for lora in self.clip.loras]
+        # loras = [(context.models.get(**lora.dict(exclude={"weight"})).context.model, lora.weight) for lora in self.clip.loras]
```
seems like this could just be removed, but I respect even updating the comments

psychedelicious replied:
Heh, I think that was a find/replace

brandonrising commented on invokeai/app/services/shared/invocation_context.py:
```diff
@@ -0,0 +1,625 @@
+            # We don't want nodes doing this, so we make a frozen copy.
+
+            config = services.configuration.get_config()
+            # TODO(psyche): If config cannot be changed at runtime, should we cache this?
```
Cache between nodes of a single graph, or between graphs?

brandonrising replied:
I think it makes sense to cache for the life of a graph

psychedelicious replied:
Because the processor operates on individual nodes, and the context is created anew for each node, I don't think we can easily cache it for the life of a graph. We could either cache it for the lifetime of the app, or an individual node.

psychedelicious replied:
I've set it up to cache the config. The config is read just-in-time, so we aren't unnecessarily reading the config for every node when that's not needed.

psychedelicious replied:
Turns out the config class already caches the config.

brandonrising commented on invokeai/app/services/shared/invocation_context.py:
```diff
@@ -0,0 +1,625 @@
+        context_data: InvocationContextData,
+    ) -> None:
+        # TODO(psyche): We are (ab)using the latents storage service as a general pickle storage
+        # service, but it is typed to work with Tensors only. We have to fudge the types here.
```
This may be a good place to break it out into separate storage services

psychedelicious replied:
Turns out we can use `ItemStorageABC` for this with a super tiny change it it. I've implemented this with support for custom save and load functions (defaulting to `torch.save` and `torch.load`).

psychedelicious replied:
I've split the services up

brandonrising commented on invokeai/app/invocations/primitives.py:
```diff
@@ -241,6 +236,14 @@ class ImageOutput(BaseInvocationOutput):
     height: int = OutputField(description="The height of the image in pixels")
 
+    @classmethod
+    def build(cls, image_dto: ImageDTO) -> "ImageOutput":
```
big fan of these build methods

brandonrising commented on invokeai/app/invocations/noise.py:
```diff
@@ -69,13 +66,13 @@ class NoiseOutput(BaseInvocationOutput):
+    def build(cls, latents_name: str, latents: torch.Tensor, seed: int) -> "NoiseOutput":
+        return cls(
+            noise=LatentsField(latents_name=latents_name, seed=seed),
+            width=latents.size()[3] * 8,
```
May make sense to make a downsampling factor optional param here in case a node has latents made with a different one ```downsampling_factor: int = 8```

psychedelicious replied:
Yeah, good call.

psychedelicious replied:
On second thought, I think adding an additional arg that is never used. I'd prefer to wait for a need for this to add that functionality.

RyanJDick replied:
I was also made uncomfortable by this at one point and created this constant to ease my discomfort: https://github.com/invoke-ai/InvokeAI/blob/f64fc2c8b76985c68e6ecba7872020e53d51a6fa/invokeai/app/invocations/latent.py#L82-L86

Feel free to use it if you want. But also fine if you leave it since this is not the focus of this PR.

psychedelicious replied:
I've added the constant, thanks. If we need to make it dynamic in the future we can always update the API - wanna keep the method signatures as minimal as possible.

psychedelicious replied:
Well, that creates a circular reference, reverted. I've made a ticket to remind me to follow up.

RyanJDick commented on invokeai/app/services/shared/invocation_context.py:
```diff
@@ -0,0 +1,625 @@
+        self.update = update
+
+
+class LatentsInterface:
```
I can't think of a reason for `LatentsInterface` to be part of the API if we were designing it from scratch today. Valid reasons to keep it are 1) it's too much work to move away from it, or 2) the migration would be harder for node developers.

To be explicit about it, why are we including it?

psychedelicious replied:
1. > it's too much work to move away from it

I think yes, but I should be accurate and instead say I don't know how much work it would take to move away from it.

```py
@invocation_output("latents_output")
class LatentsOutput(BaseInvocationOutput):
    """Base class for nodes that output a single latents tensor"""

    tensor: torch.Tensor = PrivateAttr() # should this be omitted from the graph or held in memory throughout a graph's lifetime?
    # ... the rest of the class, but modified to not need a `LatentsField` or latents name on instantiation, somehow...

# in BaseInvocation.invoke_internal
output = self.invoke(context)
if isinstance(output, LatentsOutput):
    context.services.latents.save("some_id", output.tensor)
return output
```

Something like this is simple for an output class with a single `LatentsField`, but a bit more involved for custom outputs, which can be any arbitrary pydantic model.

We'd have to traverse the output inside `invoke_internal`, pulling out all `LatentsOutput` objects to save them.

I have a feeling there is additional complexity here that I'm missing.

2. > the migration would be harder for node developers

I'm sure we do could it in such a way that it isn't difficult to migrate, so I don't think this is a good reason.

3. We are now adding special handling for tensors. It feels wrong to do it only for tensors and not also for images. What about future non-serializable or large output types - say, a 3D model?

4. We already have a pattern for _loading_ tensors. It's awkward for one side of the API to be imperative and the other not. If we save them automatically, we should load them automatically, too. I think that would require digging deeper into `Graph`, specifically the `prepare` logic.

---

I'd like to explore this further but it may make more sense to do this in the context of a deeper rewrite.

RyanJDick replied:
Sounds good to me.

When we do get around to it, I was imagining that all node outputs would just have to be serializable and we would handle them all in pretty much the same way. Some types might need custom serializers, but beyond that there would be no need for special handling of tensors.

RyanJDick commented on invokeai/app/services/shared/invocation_context.py:
```diff
@@ -0,0 +1,625 @@
+        self.get = get
+
+
+class ConditioningInterface:
```
Same question as for LatentsInterface: Why are we including this in the public API?

psychedelicious replied:
Responded above

RyanJDick reviewed (commented):
Reviewed invocation_context.py

psychedelicious commented on invokeai/app/services/shared/invocation_context.py:
```diff
@@ -0,0 +1,625 @@
+            """
+            return services.images.get_dto(image_name)
+
+        def update(
+            image_name: str,
+            board_id: Optional[str] = None,
+            is_intermediate: Optional[bool] = False,
+        ) -> ImageDTO:
```
The method exists here because I was trying to dogfood the this new context _and_ still hide the full-fat services.

If we `A)` don't try to hide the services and/or `B)` add the `WithBoard` class and handling described below, this can be removed.

## Context

`LinearUIOutputInvocation` is used only by the linear UI, and is hidden from the nodes menu in the workflows UI. It adds images to boards and changes the `is_intermediate` status for already-saved images without needing to write the image a second time.

Why?

The VAE decode node doesn't have a `board_id` field, so its outputs are saved without a board association. To auto-save images to a board, we need to take some action after the VAE decode. We are on the third solution now.

### Three solutions

- First, we waited until an image was received, then hit the API to add that image to the board. This added some jank as images appeared first in the uncategorized area, then were moved to a board after a moment.
- Second, we added a `SaveImage` node that added the image to the board, but also re-saved the image. Graphs that used the node ended up with duplicate output images (only one of which was in the gallery). This was not popular.
- Third, `LinearUIOutputInvocation` is the same as `SaveImage`, but without writing the image a second time.

The intention with these solutions was to avoid adding a board field to every image-outputting node. I perceived this as tedious to maintain, and would add more complexity to the UI for an individual node.

### Proposal

At this point, though, I think it makes sense to add a `WithBoard` class that works like `WithMetadata`. When a node inherits from `WithMetadata`, it get the metadata field "for free", without needing to worry about handling the metadata unless it's needed.

Similarly, `WithBoard` would add a board dropdown to the node, and the context wrapper would handle grabbing the board selection and saving the image with it. The node author would get board-adding functionality by just inheriting, without any changes to their node logic.

I hadn't figured out this pattern at the time I did the other solutions.

psychedelicious replied:
I've gone ahead and implemented `WithBoard` as described.

psychedelicious replied:
There are a chunk of UI changes, which are only there to support this new way to save boards. Sorry it increased the PR size a bit.

psychedelicious commented on invokeai/app/services/shared/invocation_context.py:
```diff
@@ -0,0 +1,625 @@
+            :param board_name: The name of the board to create.
+            """
+            return services.boards.create(board_name)
+
+        def get_dto(board_id: str) -> BoardDTO:
+            """
+            Gets a board DTO.
+
+            :param board_id: The ID of the board to get.
+            """
+            return services.boards.get_dto(board_id)
+
+        def get_all() -> list[BoardDTO]:
+            """
+            Gets all boards.
+            """
+            return services.boards.get_all()
+
+        def add_image_to_board(board_id: str, image_name: str) -> None:
+            """
+            Adds an image to a board.
+
+            :param board_id: The ID of the board to add the image to.
+            :param image_name: The name of the image to add to the board.
+            """
+            services.board_images.add_image_to_board(board_id, image_name)
+
+        def get_all_image_names_for_board(board_id: str) -> list[str]:
+            """
+            Gets all image names for a board.
+
+            :param board_id: The ID of the board to get the image names for.
+            """
+            return services.board_images.get_all_board_image_names_for_board(board_id)
+
+        self.create = create
+        self.get_dto = get_dto
+        self.get_all = get_all
+        self.add_image_to_board = add_image_to_board
+        self.get_all_image_names_for_board = get_all_image_names_for_board
```
Yes, to obfuscate it and reduce footgun potential. We give the keys to the castle to node authors, I'm wary of somebody writing an unintentionally dangerous node.

I know this kind of "security" isn't really possible with python - plus, you can always just import `ApiDependencies` from anywhere.

Maybe a private variable is good enough for the purposes of reducing footgun potential.

psychedelicious replied:
Ok, I've made it a private variable instead.

psychedelicious commented on invokeai/app/services/shared/invocation_context.py:
```diff
@@ -0,0 +1,625 @@
+        """Provides utility methods."""
+        self.boards = boards
+        """Provides methods to interact with boards."""
+        self.data = data
```
I'm aware of some nodes that want to grab the graph ID or workflow. How about we make it private? I don't think there's any harm in exposing it, at least not like there is by exposing the services.

psychedelicious replied:
I've made it private.

psychedelicious commented on invokeai/app/services/shared/invocation_context.py:
```diff
@@ -0,0 +1,625 @@
+        self.update = update
+
+
+class LatentsInterface:
```
Sure.

psychedelicious replied:
Ok, I've removed all (? think I got em all) references to "latents" within the services. See my latest commit on the whole PR for detail.

I don't want to change the invocation definitions or output classes though, because that would a massively breaking change.

psychedelicious commented:
@RyanJDick @brandonrising 
Regarding backwards compatibility, I unfortunately had to move a lot of files around due to circular references. The previous module paths for our exports are no longer valid, and node authors will need to change their imports to come from `invokeai.invocation_api`.

If node authors are already going to be forced to do this, I don't thing there's a huge need to retain the old context API, marked as deprecated. They will need to go change stuff anyways.

So, with this in mind and Ryan's suggestion to _not_ attempt to hide away the old `services` object, I propose:
- We provide access the old `context.services` object, but as `context._services` instead.
- We do not provide backcompat for the old properties like `context.graph_execution_state_id`, instead they will be available on `context._data`. Note that this is `context.data` in the PR at this moment, I'll change it.

The general migration path for node authors will be:
- Change your imports to all come from `invokeai.invocation_api`. Any other import is considered unstable.
- If you want to use the new, more ergonomic API, go ahead and use it.
- If you want to continue using the old API, you can change your calls from `context.services.whatever` to `context._services.whatever`

psychedelicious commented:
Note to self: Ensure all node versions are bumped correctly before merge!

psychedelicious commented:
Previously, we used `LatentsStorage` to store both "latents" tensors and conditioning data (which is a dataclass that has a tensor in it). There was a disk implementation and a forward cache wrapper.

I've generalized this into `ObjectSerializer`. I tried `PickleStorage`, which, while much more exciting of a name, seemed a bit off. The purpose of the class is to serialize and store arbitrary python objects - landed on `ObjectSerializer`. Happy to change it for a better name.

Anyways, it is a generic version of `LatentsStorage`. The `latents` service is no more - there are now `tensors` and `conditioning` services (again, I' happy to change the names).

The disk implementation is `ObjectSerializerEphemeralDisk`, named as such because it stores the tensors and conditioning object disk, and clears them on app startup. `ObjectSerializerForwardCache` is very similar to the previous class. I've added tests for both these classes.

Note: There are a few commits in the history where I though that the existing `ItemStorage` classes would work for tensors and conditioning and I implemented the ephemeral disk store. It doesn't work, though - `ItemStorage` is intended for mutable objects and idempotent item ID generation logic. For tensors, we want immutable storage and unique IDs for every object.

RyanJDick commented:
> @RyanJDick @brandonrising Regarding backwards compatibility, I unfortunately had to move a lot of files around due to circular references. The previous module paths for our exports are no longer valid, and node authors will need to change their imports to come from `invokeai.invocation_api`.
> 
> If node authors are already going to be forced to do this, I don't thing there's a huge need to retain the old context API, marked as deprecated. They will need to go change stuff anyways.
> 
> So, with this in mind and Ryan's suggestion to _not_ attempt to hide away the old `services` object, I propose:
> 
> * We provide access the old `context.services` object, but as `context._services` instead.
> * We do not provide backcompat for the old properties like `context.graph_execution_state_id`, instead they will be available on `context._data`. ~Note that this is `context.data` in the PR at this moment, I'll change it.~
> 
> The general migration path for node authors will be:
> 
> * Change your imports to all come from `invokeai.invocation_api`. Any other import is considered unstable.
> * If you want to use the new, more ergonomic API, go ahead and use it.
> * If you want to continue using the old API, you can change your calls from `context.services.whatever` to `context._services.whatever`

This makes sense to me. Though we should strongly discourage the use of `context._services`. Once it is not part of the public API we should be free to change it without even a mention in the release notes.

RyanJDick reviewed (commented):
Finished another pass of invocation_context.py. I think it's pretty close now. And I'm expecting to have very few comments on the rest of the diff 🤞

psychedelicious commented on invokeai/app/services/shared/invocation_context.py:
```diff
@@ -0,0 +1,524 @@
+        # with the execution of a specific node.
+        #
+        # While this is available within the node, it's tedious to need to pass it in on every
+        # call. We can avoid that by wrapping the method here.
```
Yes, but comment doesn't make sense without visibility into the previous behaviour. We were passing the old context in from the node to the model manager, which used both the services and data from the old context to emit events. Cleaning up the comment.

psychedelicious commented on invokeai/app/services/shared/invocation_context.py:
```diff
@@ -0,0 +1,524 @@
+            you want to override or provide a board manually!**
+        :param image_category: The category of the image. Only the GENERAL category is added \
+            to the gallery.
+        :param metadata: The metadata to save with the image, if it should have any. If the \
+            invocation inherits from `WithMetadata`, that metadata will be used automatically. \
+            **Use this only if you want to override or provide metadata manually!**
```
Good catch, my logic is backwards here. If the user provides `metadata`, we should use that and ignore the values provided by `WithMetadata`. I've fixed this.

Yes, there are community nodes that use this. They construct metadata based on their internal state and inputs.

psychedelicious commented on invokeai/app/services/shared/invocation_context.py:
```diff
@@ -0,0 +1,524 @@
+        If the current queue item has a workflow or metadata, it is automatically saved with the image.
+
+        :param image: The image to save, as a PIL image.
+        :param board_id: The board ID to add the image to, if it should be added. It the invocation \
+            inherits from `WithBoard`, that board will be used automatically. **Use this only if \
+            you want to override or provide a board manually!**
```
Thanks, fixed per other comment

psychedelicious commented on invokeai/app/services/shared/invocation_context.py:
```diff
@@ -0,0 +1,524 @@
+Many of the wrappers have the same signature as the methods they wrap. This allows us to write
+user-facing docstrings and not need to go and update the internal services to match.
+
+Note: The docstrings are in weird places, but that's where they must be to get IDEs to see them.
```
Good stuff, thanks for that info!

psychedelicious commented:
I've made the delete on startup behaviour of the new object serializer class opt-in and updated the app & tests accordingly.

RyanJDick commented on invokeai/app/services/object_serializer/object_serializer_disk.py:
```diff
@@ -0,0 +1,95 @@
+
+        deleted_count = 0
+        freed_space = 0
+        for file in Path(self._output_dir).glob("*"):
```
This `glob("*")` makes me slightly nervous. Maybe I'm just being paranoid... but I'd hate for someone to accidentally wipe a directory by passing in the wrong `output_dir`.

Some options I can think of to reduce risk:
1. Use a temp directory. I saw your comment above, but I think temp directories should _almost_ always be cleaned up correctly. The only exception I can think of is if a daemon thread is killed without proper cleanup.
2. Always use an output_dir in the system's temp dir. Before deleting, check to make sure that the path is in fact in the temp dir.
3. Use a file extension, and check that we are only deleting files with that file extension.

psychedelicious replied:
Ok, I've figured out a solution with `tempfile.TemporaryDirectory`. From the commit message:

Replace `delete_on_startup: bool` & associated logic with `ephemeral: bool` and `TemporaryDirectory`.

The temp dir is created inside of `output_dir`. For example, if `output_dir` is `invokeai/outputs/tensors/`, then the temp dir might be `invokeai/outputs/tensors/tmpvj35ht7b/`.

The temp dir is cleaned up when the service is stopped, or when it is GC'd if not properly stopped.

In the event of a catastrophic crash where the temp files are not cleaned up, the user can delete the tempdir themselves.

This situation may not occur in normal use, but if you kill the process, python cannot clean up the temp dir itself. This includes running the app in a debugger and killing the debugger process - something I do relatively often.

Tests updated.

RyanJDick reviewed (commented):
Finished reviewing all of the files today. I only skimmed the frontend files - I'm not familiar enough to review with any rigor.

Just minor comments left.

psychedelicious commented on invokeai/app/services/object_serializer/object_serializer_disk.py:
```diff
@@ -0,0 +1,95 @@
+        self.__obj_class_name: Optional[str] = None
+
+    def start(self, invoker: "Invoker") -> None:
+        self._invoker = invoker
```
Yep, leftover from previous version where it was necessary. Removed

psychedelicious commented on invokeai/app/services/object_serializer/object_serializer_forward_cache.py:
```diff
@@ -0,0 +1,62 @@
+        self._cache_ids = Queue[str]()
+        self._max_cache_size = max_cache_size
+
+    def start(self, invoker: "Invoker") -> None:
+        self._invoker = invoker
+        start_op = getattr(self._underlying_storage, "start", None)
+        if callable(start_op):
+            start_op(invoker)
+
+    def stop(self, invoker: "Invoker") -> None:
+        self._invoker = invoker
+        stop_op = getattr(self._underlying_storage, "stop", None)
+        if callable(stop_op):
+            stop_op(invoker)
```
My understanding - the original intent is twofold:
1. Make things services that can have startup and shutdown logic.
2. Provide access to other services via the start method, instead of needing to provide all deps explicitly in the `__init__` method.

I'm not familiar enough with architectural design to comment on whether this pattern is good or bad

psychedelicious commented on pyproject.toml:
```diff
@@ -279,3 +281,38 @@ module = [
 ]
 #=== End: MyPy
+
+[tool.pyright]
```
Yeah, they have the same goals - static type analysis. I'm firmly on team pyright now, though I don't think I really use either tool to its fullest. I don't use mypy any more. 

Pyright wins:
- The VSCode python LSP (pylance) is built on pyright. In my experience, it's faster, less buggy, and better-integrated than the mypy extension / daemon. Plus, you don't need the extra tool. 
- Pyright seems to do better in many situations. Most notably, pyright handles narrowing and type guards more adeptly. It widens untyped variables as you assign to them, which is nice when working with untyped external code.
- Maintainer is very responsive. Project backed by MS. Also - and this is only anecdotal - it is written in typescript, and I've found TS devs to be _far_ more pedantic about their tooling than python devs.

Mypy wins:
- Mypy seems to be more configurable but maybe I just haven't dug into to pyright enough yet

I've been adding this relatively strict config when I need to edit python code, then remove it when done. VSCode will use the config and you'll be assaulted by red squigglies with it. I'll remove it before merging.

psychedelicious commented on invokeai/app/services/shared/invocation_context.py:
```diff
@@ -0,0 +1,520 @@
+        )
+
+
+deprecation_version = "3.7.0"
+removed_version = "3.8.0"
```
Ah, we haven't talked through this. If we include split of `services.latents` into `services.[tensors|conditioning]`, then we'll need to add another layer of backcompat to map the old service to these two new ones. Part of me thinks it's not worth handling deprecation here.

The values and messages are placeholders until we decide on how to handle this.

psychedelicious replied:
See https://github.com/invoke-ai/InvokeAI/pull/5491#issuecomment-1936952131 for the broader discussion on deprecation

psychedelicious commented:
Barring any concerns with the implementation of a temp dir for `ObjectSerializerDisk`, the last remaining discussion items are around backwards compatibility / deprecation. 

## 1. Partial backwards compatibility at best

While implementing this PR, I ran into circular dependency issues. These couldn't be fixed with `if TYPE_CHECKING`, and required folder structure changes.

As a result, **the release that includes the new API is a breaking change, and all custom nodes _must_ change their import to come from `invokeai.invocation_api`**. I don't think there is any remotely reasonable / maintainable solution.

This PR includes a backwards compatibility for the `context` object, but we cannot reasonably provide backcompat for the import paths.

I question releasing with _partial_ backwards compatibility. It sends the message that it's OK for node authors to change only the import paths and call it a day. When we eventually remove the back compat layer, they'll need to go back and make another round of changes again. This situation is untenable.

I think we have to bite the bullet here and make it a clean break - no backwards compatibility at all. This probably means the next release is a major release - Invoke v4.0.0.

> Note: This PR removes `services.latents` entirely. I've written a simple backcompat layer for this locally that seems reasonable, in case we do decide to release with partial back compat.

## 2. Upcoming Model Manager v2

We have an opportunity to insulate node authors from the upcoming model manager changes. Proposal:

- Methods in `ModelsInterface` are namespaced or suffixed e.g. `context.models.v1.load` or `context.models.load_v1`.
- When the new model manager is released, we add new methods with a `v2` namespace.
- We ensure the `v1` methods still function until the next major release.

hipsterusername commented:
Big break makes sense. We just need to offer some assurances in changes going forward and make it clear what needs to change to fix issues.

I think a 4.0.0 release with some of the new things we’ve been working on is reasonable.

lstein commented:
> ## 2. Upcoming Model Manager v2
> We have an opportunity to insulate node authors from the upcoming model manager changes. Proposal:
> 
> * Methods in `ModelsInterface` are namespaced or suffixed e.g. `context.models.v1.load` or `context.models.load_v1`.
> * When the new model manager is released, we add new methods with a `v2` namespace.
> * We ensure the `v1` methods still function until the next major release.

I've looked through the proposed model manager wrapper API and I question whether we need to maintain two versions of the API. All internal nodes, and I would hazard all contributed nodes, load models through the `ModelInfo` (renamed `LoadedModelInfo`) node class. In the current implementation, this class contains attributes for the model's name, base, type, and submodel. In MMv2, this class contains the model key and submodel only. Once the frontend switches over to using the MMv2 `LoadedModelInfo` class, the invocation services wrapper will start using the key-style identification and the downstream nodes won't see any change at all.

There are more significant differences in the model installation, but this functionality doesn't seem to be part of the wrapper at present.

psychedelicious commented:
Superseded by `next` branch
